### PR TITLE
add re3 documentation

### PR DESCRIPTION
**Pull request by n30111**
This PR adds documentation for RE3.

### CODE REVIEW

File: doc/source/rllib-algorithms.rst [hunk 1]
```diff
@@ -887,3 +887,61 @@ Intrinsic rewards for each env-step are calculated by taking the euclidian dista
...
+	 "rho": 0.1, # Beta decay factor, used for on-policy algorithm.
+         "k_nn": 50, # Number of neighbours to set for K-NN entropy estimation.
+         # Configuration for the encoder network, producing embedding vectors from observations.
+         # This can be used to configure fcnet- or conv_net setups to properly process any
+         # observation space. By default uses the Policy model configuration.
+         "encoder_net_config": {
+             "fcnet_hiddens": [],
+             "fcnet_activation": "relu",
+         },
+         "beta": 0.2 # Hyperparameter to choose between exploration and exploitation.
```
Comment (jbedorf) [inline doc/source/rllib-algorithms.rst]: Maybe this explanation can be a bit longer. As in what does 0.2 mean? Is that a chance of 1 in 5 on exploitation?

File: doc/source/rllib-algorithms.rst [hunk 2]
```diff
@@ -887,3 +887,61 @@ Intrinsic rewards for each env-step are calculated by taking the euclidian dista
...
+         # Schedule to use for beta decay, one of constant" or "linear_decay".
```
Comment (jbedorf) [inline doc/source/rllib-algorithms.rst:933]:
```suggestion
         # Schedule to use for beta decay, one of "constant" or "linear_decay".
```

File: doc/source/rllib-algorithms.rst [hunk 3]
```diff
@@ -887,3 +887,61 @@ Intrinsic rewards for each env-step are calculated by taking the euclidian dista
...
+         # exploration, e.g. EpsilonGreedy for DQN, StochasticSampling for PG/SAC).
+         "sub_exploration": {
+             "type": "StochasticSampling",
+         }
+    }
+
+
+**Functionality**
+RLlib's RE3 is based on `"Random Encoders for Efficient Exploration" described in this paper here <https://arxiv.org/pdf/2102.09430.pdf>`__.
+RE3 quantifies exploration based on state entropy. The entropy of a state is calculated based on its distance from K nearest neighbors states present in the replay buffer (we use train batch in this implementation) in the latent space.
```
Comment (jbedorf) [inline doc/source/rllib-algorithms.rst]:
```suggestion
RE3 quantifies exploration based on state entropy. The entropy of a state is calculated based on its distance from K nearest neighbors states present in the replay buffer in the latent space (train batch is used in this implementation). 
```

File: doc/source/rllib-algorithms.rst [hunk 4]
```diff
@@ -887,3 +887,61 @@ Intrinsic rewards for each env-step are calculated by taking the euclidian dista
...
+             "type": "StochasticSampling",
+         }
+    }
+
+
+**Functionality**
+RLlib's RE3 is based on `"Random Encoders for Efficient Exploration" described in this paper here <https://arxiv.org/pdf/2102.09430.pdf>`__.
+RE3 quantifies exploration based on state entropy. The entropy of a state is calculated based on its distance from K nearest neighbors states present in the replay buffer (we use train batch in this implementation) in the latent space. 
+The entropy of the state is considered as intrinsic rewards and added to the extrinsic rewards for policy optimization when external rewards are available, else used as "intrinsic rewards" for unsupervised pre-training of RL agent. RE3 allows agents to learn in sparse-reward or even no-reward environments by
+considering the state entropy as "intrinsic rewards".
```
Comment (jbedorf) [inline doc/source/rllib-algorithms.rst]:
```suggestion
The state entropy is considered as an intrinsic reward and for policy optimization added to the extrinsic reward when available.  If the extrinsic reward is not available then the state entropy is used as "intrinsic reward" for unsupervised pre-training of the RL agent. 
RE3 further allows agents to learn in sparse-reward or even no-reward environments by
using the state entropy as "intrinsic rewards".
```

Please verify if this rewrite matches the original intent.

**Review by jbedorf — COMMENTED** [review]
Looks good, I could follow the description. Just some minor changes and questions.

File: doc/source/rllib-algorithms.rst (hunk 1 above)
Comment (n30111) [reply doc/source/rllib-algorithms.rst]: Beta weights the intrinsic rewards, beta 1 means intrinsic rewards added as it is. `reward = r + beta * intrinsic_rewards`

Comment (n30111) [reply doc/source/rllib-algorithms.rst]: I have updated the doc string, please check.

File: doc/source/rllib-algorithms.rst [hunk 5]
```diff
@@ -887,3 +887,65 @@ Intrinsic rewards for each env-step are calculated by taking the euclidian dista
...
+    config = sac.DEFAULT_CONFIG.copy()
+    config["env"] = "Pendulum-v1"
+    config["seed"] = 12345
+    config["callbacks"] = RE3Callbacks
+    config["exploration_config"] = {
+    	"type": "RE3",
+         # the dimensionality of the observation embedding vectors in latent space.
+         "embeds_dim": 128, 
+	 "rho": 0.1, # Beta decay factor, used for on-policy algorithm.
+         "k_nn": 50, # Number of neighbours to set for K-NN entropy estimation.
```
Comment (jbedorf) [inline doc/source/rllib-algorithms.rst]:
```suggestion
         # the dimensionality of the observation embedding vectors in latent space.
         "embeds_dim": 128, 
         "rho": 0.1, # Beta decay factor, used for on-policy algorithm.
         "k_nn": 50, # Number of neighbours to set for K-NN entropy estimation.
```
Tab/space change

File: doc/source/rllib-algorithms.rst [hunk 6]
```diff
@@ -887,3 +887,65 @@ Intrinsic rewards for each env-step are calculated by taking the euclidian dista
...
+         # Configuration for the encoder network, producing embedding vectors from observations.
+         # This can be used to configure fcnet- or conv_net setups to properly process any
+         # observation space. By default uses the Policy model configuration.
+         "encoder_net_config": {
+             "fcnet_hiddens": [],
+             "fcnet_activation": "relu",
+         },
+         # Hyperparameter to choose between exploration and exploitation. A higher value of beta add
```
Comment (jbedorf) [inline doc/source/rllib-algorithms.rst]:
```suggestion
         # Hyperparameter to choose between exploration and exploitation. A higher value of beta adds
```

**Review by jbedorf — APPROVED** [review]
Changes look good, thanks!

File: doc/source/rllib-algorithms.rst (hunk 1 above)
Comment (jbedorf) [reply doc/source/rllib-algorithms.rst]: Yep, this helps. Thank you

**Pull request merged**